### PR TITLE
amadeus: Update to REV10

### PR DESCRIPTION
**Pull request by Thog**
This implements all the changes made with REV10 on 13.0.0.

### CODE REVIEW

File: Ryujinx.Audio/Renderer/Dsp/BiquadFilterHelper.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+                    state.State1 = state.State0;
+                    state.State0 = input;
+                    state.State3 = state.State2;
+                    state.State2 = output;
+
+                    outputBuffer[i] = output;
+                }
+            }
+
```
Comment (AcK77) [inline Ryujinx.Audio/Renderer/Dsp/BiquadFilterHelper.cs:98]: Extra empty line

**Review by AcK77 — APPROVED** [review]

File: Ryujinx.Audio/Renderer/Server/CommandGenerator.cs [hunk 2]
```diff
@@ -556,6 +568,52 @@ private void GenerateLimiterEffect(uint bufferOffset, LimiterEffect effect, int
             }
         }
 
+        private void GenerateCaptureEffect(uint bufferOffset, CaptureBufferEffect effect, int nodeId)
+        {
+            Debug.Assert(effect.Type == EffectType.CaptureBuffer);
+
+            if (effect.IsEnabled)
+            {
+                effect.GetWorkBuffer(0);
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderer/Server/CommandGenerator.cs:577]: Is the return value really supposed to be unused?

File: Ryujinx.Audio/Renderer/Server/CommandGenerator.cs [hunk 3]
```diff
@@ -556,6 +568,52 @@ private void GenerateLimiterEffect(uint bufferOffset, LimiterEffect effect, int
...
+                    {
+                        updateCount = 0;
+                    }
+                    else
+                    {
+                        updateCount = newUpdateCount;
+                    }
+
+                    _commandBuffer.GenerateCaptureEffect(bufferOffset,
+                                                     effect.Parameter.Input[i],
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderer/Server/CommandGenerator.cs:601]: Alignment is off.

File: Ryujinx.Audio/Renderer/Server/CommandProcessingTimeEstimatorVersion4.cs [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+using System;
+using System.Diagnostics;
+using static Ryujinx.Audio.Renderer.Parameter.VoiceInParameter;
+
+namespace Ryujinx.Audio.Renderer.Server
+{
+    /// <summary>
+    /// <see cref="ICommandProcessingTimeEstimator"/> version 4. (added with REV10)
+    /// </summary>
+    public class CommandProcessingTimeEstimatorVersion4 : ICommandProcessingTimeEstimator
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderer/Server/CommandProcessingTimeEstimatorVersion4.cs:30]: Why not derive this class from the version 3 class (and cascade in a similar way for previous versions) to avoid all the code duplication? The new functions would be made virtual and overriden on this class.

File: Ryujinx.HLE/HOS/Services/Audio/AudioRenderer/AudioDeviceServer.cs [hunk 5]
```diff
@@ -38,8 +38,6 @@ public ResultCode ListAudioDeviceName(ServiceCtx context)
 
                 if ((position - basePosition) + (ulong)buffer.Length > size)
                 {
-                    Logger.Error?.Print(LogClass.ServiceAudio, $"Output buffer size {size} too small!");
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Audio/AudioRenderer/AudioDeviceServer.cs:41]: This is not related to the REV10 update, would be nice to get an explanation of why it was removed.

File: Ryujinx.Audio/Renderer/Server/CommandGenerator.cs [hunk 6]
```diff
@@ -556,6 +568,52 @@ private void GenerateLimiterEffect(uint bufferOffset, LimiterEffect effect, int
...
+                int i = 0;
+                uint writeOffset = 0;
+
+                for (uint channelIndex = effect.Parameter.ChannelCount; channelIndex != 0; channelIndex--)
+                {
+                    uint newUpdateCount = writeOffset + _commandBuffer.CommandList.SampleCount;
+
+                    uint updateCount;
+
+                    if ((channelIndex - 1) != 0)
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderer/Server/CommandGenerator.cs:591]: Why not `channelIndex != 1`?

File: Ryujinx.Audio/Renderer/Dsp/BiquadFilterHelper.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        private const int FixedPointPrecisionForParameter = 14;
+
+        /// <summary>
+        /// Apply a single biquad filter.
+        /// </summary>
+        /// <remarks>This is implemented with a direct form 2.</remarks>
+        /// <param name="parameter">The biquad filter parameter</param>
+        /// <param name="state">The biquad filter state</param>
+        /// <param name="outputBuffer">The output buffer to write the result.</param>
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.Audio/Renderer/Dsp/BiquadFilterHelper.cs:35]: Or add a `.` on the other ones I guess. Most of them seems to have it.

File: Ryujinx.Audio/Renderer/Server/CommandGenerator.cs (hunk 2 above)
Comment (Thog) [reply Ryujinx.Audio/Renderer/Server/CommandGenerator.cs:577]: Yes and Nintendo does that too (it's to pin the reference)

File: Ryujinx.Audio/Renderer/Dsp/Command/GroupedBiquadFilterCommand.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+        public ulong EstimatedProcessingTime { get; set; }
+
+        private BiquadFilterParameter[] _parameters;
+        private Memory<BiquadFilterState> _biquadFilterStates;
+        private int _inputBufferIndex;
+        private int _outputBufferIndex;
+        private bool[] _isInitlaized;
+
+        public GroupedBiquadFilterCommand(int baseIndex, ref ReadOnlySpan<BiquadFilterParameter> filters, Memory<BiquadFilterState> biquadFilterStateMemory, int inputBufferOffset, int outputBufferOffset, ReadOnlySpan<bool> isInitialized, int nodeId)
```
[the remark this answers is not in the document]
Comment (Thog) [reply Ryujinx.Audio/Renderer/Dsp/Command/GroupedBiquadFilterCommand.cs:40]: missed to fix that oops

File: Ryujinx.Audio/Renderer/Dsp/State/AuxiliaryBufferHeader.cs [hunk 9]
```diff
@@ -20,18 +20,22 @@
...
+            private const uint TotalSampleCountPosition = 0xC;
 
             public uint ReadOffset;
             public uint WriteOffset;
-            private uint _reserved;
+            public uint LostSampleCount;
+            public uint TotalSampleCount;
+            private unsafe fixed uint _unknown[12];
 
             public static uint GetReadOffset(IVirtualMemoryManager manager, ulong bufferAddress)
```
[the remark this answers is not in the document]
Comment (Thog) [reply Ryujinx.Audio/Renderer/Dsp/State/AuxiliaryBufferHeader.cs:40]: Problem is that some fields could be updated concurrently on the "CPU" side and ending up being overwritten by the DSP side. This was the solution that was used on initial implementation as a result

[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.Audio/Renderer/Dsp/State/AuxiliaryBufferHeader.cs:40]: Oh yea, that is a problem for the Set* methods.

File: Ryujinx.HLE/HOS/Services/Audio/AudioRenderer/AudioDeviceServer.cs (hunk 5 above)
Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Audio/AudioRenderer/AudioDeviceServer.cs:41]: Well this error should have been removed a long time ago, this is a normal behavior that game have

**Review by gdkchan — APPROVED** [review]
lgtm, but I don't have the means to test the new code right now, as afaik nothing uses it yet.
